### PR TITLE
[ops] feat: veomni_fa_function

### PR DESCRIPTION
**Pull request by Coach257**
### What does this PR do?

1. patch `flash_attention_2` and `flash_attention_3` in `ALL_ATTENTION_FUNCTIONS`
2. define new mask_func in `ALL_MASK_ATTENTION_FUNCTIONS`
3. add external fa ops in `ALL_FLASH_ATTENTION_FUNCTIONS`, using `transformers.modeling_flash_attention_utils._flash_attention_forward` for default
    - delete seed_kernels related, for internal use only

4. `flash_attention_forward` up to date (transformers==4.57.3)
5. models support fa3 now: qwen2, qwen3, qwen3moe

### CODE REVIEW

**Comment by Luosuu** [comment]
Hi, let's coordinate on this since I am support fa3 now #305

File: veomni/models/transformers/deepseek_v3/modeling_deepseek.py [hunk 1]
```diff
@@ -53,6 +53,7 @@
 
 from ....ops import causallm_loss_function, fused_moe_forward
 from ....utils.import_utils import is_liger_kernel_available
+from ...modeling_utils import ALL_ATTENTION_FUNCTIONS
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek.py]: wondering if we can keep the current impl that just register the transformers ALL_ATTENTION_FUNCTIONS? which requires less modifications to the modeling code.

File: veomni/models/modeling_utils.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+logger = logging.get_logger(__name__)
+
+# Global AttentionInterface shared by all models which do not need to overwrite any of the existing ones
+ALL_ATTENTION_FUNCTIONS: AttentionInterface = AttentionInterface()
+
+ALL_ATTENTION_FUNCTIONS["flash_attention_3"] = flash_attention_forward
+ALL_ATTENTION_FUNCTIONS["flash_attention_2"] = flash_attention_forward
+
+ALL_FLASH_ATTENTION_FUNCTIONS["flash_attention_2"] = _flash_attention_forward
```
Comment (piyifan123) [inline veomni/models/modeling_utils.py]: nit: a nitpick is that it's kind of confusing for users since they still confuses this "flash_attention_2" with the HF "flash_attention_2" impl. One idea is to give it another name like "veomni_ulysses_fa2".

File: veomni/models/modeling_utils.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+from ..ops.attention import ALL_FLASH_ATTENTION_FUNCTIONS, flash_attention_forward
+from ..utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+# Global AttentionInterface shared by all models which do not need to overwrite any of the existing ones
+ALL_ATTENTION_FUNCTIONS: AttentionInterface = AttentionInterface()
+
+ALL_ATTENTION_FUNCTIONS["flash_attention_3"] = flash_attention_forward
```
Comment (piyifan123) [inline veomni/models/modeling_utils.py]: i feel it's a little bit confusing for users that fa2 and fa3 shares the same interface and calling arguments. IIUC, this just uses the transformers logic that when impl=None, it just decides based on which package is installed which I think it's not very clear to users and developers. 

Can we instead bind impl=fa2 and impl=fa3 to the functions?

File: veomni/ops/attention.py [hunk 4]
```diff
@@ -28,28 +30,56 @@ def flash_attention_forward(
...
 
-    # FA2 always relies on the value set in the module, so remove it if present in kwargs to avoid passing it twice
-    kwargs.pop("is_causal", None)
+    # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+    # therefore the input hidden states gets silently casted in float32. Hence, we need
+    # cast them back in the correct dtype just to be sure everything works as expected.
+    # This might slowdown training & inference so it is recommended to not cast the LayerNorms
+    # in fp32. (usually our RMSNorm modules handle it correctly)
+    target_dtype = None
+    if query.dtype == torch.float32:
```
Comment (piyifan123) [inline veomni/ops/attention.py:64]: wondering if we can just have a PEFT specific implementation variant like "veomni_peft_fa2" so it's clear for developers and users that this logic is only triggered for peft cases.

File: veomni/models/modeling_utils.py (hunk 3 above)
Comment (Coach257) [reply veomni/models/modeling_utils.py]: fa2 and fa3 shares the same Ulysses patched `flash_attention_forward` but different `_flash_attention_forward`. The difference is controlled by the `attn_implementation`. See https://github.com/huggingface/transformers/blob/bdc85cb85c8772d37aa29ce447860b44d7fad6ef/src/transformers/modeling_flash_attention_utils.py#L608

File: veomni/ops/attention.py (hunk 4 above)
Comment (Coach257) [reply veomni/ops/attention.py:64]: It's just a copy of https://github.com/huggingface/transformers/blob/bdc85cb85c8772d37aa29ce447860b44d7fad6ef/src/transformers/integrations/flash_attention.py#L68
I will update this to 5.0.0

File: veomni/models/modeling_utils.py (hunk 2 above)
Comment (Coach257) [reply veomni/models/modeling_utils.py]: But to use `veomni_ulysses_fa2`， the user should load model by `AutoModel.from_pretrained(attn_implementation="veomni_ulysses_fa2")`. I think it's strange.

Comment (Coach257) [reply veomni/models/modeling_utils.py]: When `sp_size>1`, use `veomni_ulysses_fa2`; when `sp_size=1 and seed_kernels_availabale`, also use `veomni_ulysses_fa2`. So I think when using a veomni patched model, all the functions should be replaced with veomni patched fa forward.

File: veomni/models/transformers/deepseek_v3/modeling_deepseek.py (hunk 1 above)
Comment (Coach257) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek.py]: See https://github.com/ByteDance-Seed/VeOmni/pull/304#discussion_r2629406111

**Comment by Coach257** [comment]
> Hi, let's coordinate on this since I am support fa3 now #305

Qwen3Moe; Qwen3; Qwen2 fa3 supported and ci added @Luosuu

File: veomni/models/modeling_utils.py (hunk 3 above)
Comment (KKZ20) [reply veomni/models/modeling_utils.py]: I think it is the same implemention for ulysses?

Comment (KKZ20) [reply veomni/models/modeling_utils.py]: you can let transformers to decide attn_implementation by analyzing your env

File: veomni/models/modeling_utils.py (hunk 2 above)
Comment (Coach257) [reply veomni/models/modeling_utils.py]: renameed as veomni_flash_attention_2

File: veomni/models/transformers/deepseek_v3/modeling_deepseek.py (hunk 1 above)
Comment (Coach257) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek.py]: register to hf

File: veomni/models/transformers/llama/modeling_llama.py [hunk 5]
```diff
@@ -640,7 +640,7 @@ def _update_causal_mask(
         past_key_values: Cache,
         output_attentions: bool,
     ):
-        if self.config._attn_implementation == "flash_attention_2":
+        if "flash_attention_2" in self.config._attn_implementation:
```
Comment (Luosuu) [inline veomni/models/transformers/llama/modeling_llama.py]: do we need to handle fa3 here as well

File: veomni/models/transformers/llama/modeling_llama.py
Comment (Luosuu) [inline veomni/models/transformers/llama/modeling_llama.py:635]: I see we deleted `_update_causal_mask` for Qwen3 model to match the latest huggingface transformers. Do we still keep it for deepseek models?

File: veomni/utils/arguments.py [hunk 6]
```diff
@@ -82,9 +82,17 @@ class ModelArguments:
...
+            "sdpa",
+            "flash_attention_2",
+            "flash_attention_3",
+            "veomni_flash_attention_2",
+            "veomni_flash_attention_3",
+            "native-sparse",
+        ]
     ] = field(
-        default="flash_attention_2",
+        default="veomni_flash_attention_2",
```
Comment (Luosuu) [inline veomni/utils/arguments.py]: maybe we need to explain what is the diff between `flash_attention_2` and `veomni_flash_attention_2`. Also, does this mean all the job yamls need to be updated if they are using SP?

File: pyproject.toml [hunk 7]
```diff
@@ -67,6 +67,7 @@ gpu = [
   "nvidia-nccl-cu12",
   "liger-kernel",
   "flash-attn",
+  "flash-attn-3",
```
Comment (Luosuu) [inline pyproject.toml:70]: maybe make flash-attn-3 as a single option so that we install fa3 like `uv sync --extra gpu --extra fa3`. I am not sure whether it is fine to install fa3 for non-Hopper GPUs.

File: veomni/models/transformers/llama/modeling_llama.py
Comment (Coach257) [reply veomni/models/transformers/llama/modeling_llama.py:635]: Prioritize updating the Qwen series models first. All subsequent models will be updated to transformers 5.0.0 by applying patches to the corresponding `modeling.py`.

File: veomni/utils/arguments.py (hunk 6 above)
Comment (Coach257) [reply veomni/utils/arguments.py]: The default value for argument `attn_implementation` is now `veomni_flash_attention_2`. If users want to run **native transformers modeling**, they can set `flash_attention_2`.

File: pyproject.toml (hunk 7 above)
Comment (Coach257) [reply pyproject.toml:70]: It's runtime error. Set `veomni_flash_attention_2` instead of `fa3`.

File: veomni/models/transformers/llama/modeling_llama.py (hunk 5 above)
Comment (Coach257) [reply veomni/models/transformers/llama/modeling_llama.py]: Yes

**Comment by Luosuu** [comment]
now it looks good to me. But let @piyifan123 also take a look

File: veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py [hunk 8]
```diff
@@ -758,8 +758,14 @@ def forward(
         if position_ids is None:
             position_ids = cache_position.unsqueeze(0)
 
-        causal_mask = self._update_causal_mask(
-            attention_mask, inputs_embeds, cache_position, past_key_values, output_attentions
+        mask_function = create_causal_mask if self.config.sliding_window is None else create_sliding_window_causal_mask
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py:761]: is this following the tf 5.0 convention?

**Review by piyifan123 — APPROVED** [review]

**Review by FoolPlayer — APPROVED** [review]

File: veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py (hunk 8 above)
Comment (Coach257) [reply veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py:761]: Yes
